### ISSUE TITLE
Write Stitching.Manager Integration Tests

### ISSUE THREAD

**Issue by Zalgo2462** [activecm/ipfix-rita#12]
# Stitching.Manager Integration Tests

### Current Status
No integration tests have been written to ensure IPFIX flows are stitched together correctly. 

### Goal
Ensure `stiching.Manager` correctly stitches flows. 

#### Terminology

- **Flow**: A one-way connection from one host to another
- **Flow Key/ Aggregate Query**: The information which uniquely identifies a session at a given time. 
  - Two IP addresses (Source/ Destination, IPAddressA/ IPAddressB)
  - Two ports (Source/ Destination, PortA/ PortB)
  - The protocol 
  - The IP address of the IPFIX exporter which sent us the flows which make up the session
- **Session Aggregate**: A two-way connection from one host to another and back. Created by aggregating **Flows**
- **sameSessionTimeout**: If a second flow starts after another flow ended, it may be aggregated with it to create a session. The flows are only aggregated if the second flow began after the first flow ended and the amount of time passed between the flows is less than or equal to the **sameSessionTimeout**. 

#### Getting Started
A new integration test can be created by creating the file `manager_test.go`. Label the package as `stiching_test`. Next, create a standard go test function `TextXXX(t *testing.T) {}`. Then, start a new integration test using 
```
	env, cleanup := environmenttest.SetupIntegrationTest(t)
	defer cleanup()
```

The `env` variable contains configuration information, and most importantly, a connection to a dockerized MongodDB. 

From here, instantiate a new `stitching.Manager`, create a go routine to feed the `input` channel, and a simple `writer` implementation to retrieve the outputs from the manager. 

The `require` package is immensely helpful in go tests. See the other test files for usage.

A fake flow with random data can be generated with `ipfix.NewFlowMock()`, then the needed fields can be overwritten as needed. 

#### The Stitching Process
When a **flow** is added to a **session aggregate**, the **flow's** one-way data is added to a two-way data structure. While a **flow** has a field `OctetTotalCount` representing the amount of bytes sent from `SourceIPAddress` to `DestinationIPAddress`, a **session aggregate** has two fields `OctetTotalCountAB` and `OctetTotalCountBA` representing the amount of bytes sent between `IPAddressA` and `IPAddressB` in each direction. The true source and destination of the **session aggregate** are determined when the aggregate is ready to be written out. If `FlowStartMillisecondsAB < FlowStartMillisecondsBA`, then `IPAddressA` is considered the source and `IPAddressB` is considered the destination, otherwise `IPAddressB` is the source and `IPAddressA` is the destination. 

It is critical that **flows** are assigned to the `AB` and `BA` sections of **session aggregates** in a non-ambiguous, reproducible manner. To ensure this happens, the host (`SourceIPAddress` or `DestinationIPAddress`) which comes first alphabetically within a **flow**, will be assigned to a **session aggregate's** `IPAddressA` field, and the other host will be assigned to the `IPAddressB` field. 

##### Zero-Stitched Flows
When a **session aggregate** is first created from a single **flow**, the flow is said to be *stitched with zeroes*. Since a flow only contains one-way data, the other side of the connection cannot be known. This lack of data is represented with zero values. 

For example given the **flow**:
```
flow1 := ipfix.NewFlowMock()
//Ensure the source comes before the destination alphabetically
//to ensure the source is mapped to host "A", and the destination is
//mapped to host "B"
flow1.MockSourceIPAddress = "1.1.1.1"
flow1.MockDestinationIPAddress = "2.2.2.2"
flow1.MockProtocolIdentifier = protocols.ICMP
flow1.MockFlowEndReason = ipfix.IdleTimeout
```
we can create a **session aggregate**:
```
var sessAgg session.Aggregate
srcMapping, err := session.FromFlow(flow, &sessAgg)
if err != nil {
	return err
}
```
`srcMapping` lets us know whether the **flow** was mapped into the AB portion of the **session aggregate** or the BA portion. We can check this like so:
```
if srcMapping == session.ASource { //flow mapped into AB portion }
else { //flow mapped into BA portion }
```
Our `sessAgg` contains the following data (psuedocode):
```
sessAgg = struct {
	AggregateQuery: struct { 
		IPAddressA: "1.1.1.1"
		PortA: <random port from mock flow>

		IPAddressB:  "2.2.2.2"
		PortB: <random port from mock flow>

		ProtocolIdentifier: protocols.ICMP

		Exporter: <random ip address from mock flow>`
	}

	FlowStartMillisecondsAB:  <random time from mock flow>
	FlowEndMillisecondsAB:  <random time from mock flow>
	FlowStartMillisecondsBA: 0
	FlowEndMillisecondsBA:  0

	OctetTotalCountAB: <random amount of data from mock flow>
	OctetTotalCountBA: 0
	PacketTotalCountAB: <random amount of data from mock flow>
	PacketTotalCountBA: 0

	FlowEndReasonAB: <random reason from mock flow>
	FlowEndReasonBA: ipfix.Nil
}
```
Since all of the BA fields (`FlowStartMillisecondsBA`, `FlowEndMillisecondsBA`, `OctetTotalCountBA`, and `FlowEndReasonBA`) are zero (or `ipfix.Nil` in the case of `FlowEndReason`), the flow is said to have been *stitched with zeroes*. 

#### Sessions From Multiple Flows
Additional **flows** can be aggregated into an existing **session aggregate** provided the **flow** has a matching **Flow Key / Aggregate Query**. First, the new **flow** is mapped into either the AB or BA section of the **session aggregate**. The `OctetTotalCount` and `PacketTotalCount` of the flow are added to the corresponding fields in the matching AB/ BA section of the aggregate. If the `FlowStartMilliseconds` of the new **flow** is earlier than corresponding field in the **session aggregate** or if the field is unset in the aggregate, the value is overwritten with the **flow's**. If the `FlowEndMilliseconds` of the new **flow** is later than the corresponding field in the **session aggregate** or if the field is unset in the aggregate, the value is overwritten with the **flow's** and the `FlowEndReason` is updated as well. 

### When to Stitch Flows
Flows should only be stitched when it logically makes sense. Unfortunately, this depends on the protocol.

#### When to Stitch ICMP(v6) (and Unknown Protocol) Flows
ICMP is a connectionless protocol in which sessions are not maintained (@william-stearns can you confirm?). In this case, flows are never stitched together. They are all stitched with zeroes. 

We use this model to deal with unknown protocols as well. 

#### When to Stitch UDP
UDP is a connectionless protocol in which sessions are not maintained at the protocol level. However, UDP sessions are often maintained in the application layer unlike ICMP. Since UDP has no support for sessions at the protocol layer, the most common `FlowEndReason` for UDP flows is simply `IdleTimeout`. In order to recover the application layer sessions, we have to make intelligent guesses. 

Our guessing scheme relies on the **sameSessionTimeout**. If a **flow** from host A to host B ended at time t<sub>FlowEnd</sub>, and another flow from host A to host B or host B to host A starts at time t<sub>FlowStart</sub>, then the two **flows** will be stitched if t<sub>FlowStart</sub> <= t<sub>FlowEnd</sub> + **sameSessionTimeout**. Otherwise the two **flows** will each be stitched with zeroes. 

#### When to Stitch TCP
TCP is a connection oriented protocol, so we don't have to make as many guesses. However, TCP monitoring isn't perfect. During normal operation, TCP **flows** will have `FlowEndReason` fields marked with `EndOfFlow`. `EndOfFlow` signals one end of the TCP session was tore down cleanly. 

When the `FlowEndReason` for a given side of a **session aggregate** (AB/BA) is `EndOfFlow`, we make sure no more flows are stitched into that side of the aggregate. Otherwise, TCP is treated the same way as UDP.

### Test Cases
The above information is summarized in the following table of test cases. 
A Google Sheet with the same information can be found [here](https://docs.google.com/spreadsheets/d/13NSKgSERqTedtLl2B_pniDWdB2ztUck4iiRfI2S1Tqs/edit?usp=sharing)

| Test Name                                   | Protocol | Flow1 EndFlowReason | Flow 1 Direction | Flow 2 Direction | Flow2.FlowStartMilliseconds <= Flow1.EndMilliseconds + sameSessionTimeout | # Of Output Session Aggregates | # Of Errors | Session Aggregate 1 Results AB | Session Aggregate 1 Results BA | Session Aggregate 2 Results AB | Session Aggregate 2 Results BA | 
|---------------------------------------------|----------|---------------------|------------------|------------------|---------------------------------------------------------------------------|--------------------------------|-------------|--------------------------------|--------------------------------|--------------------------------|--------------------------------| 
| wn                                          |          |                     |                  |                  |                                                                           |                                |             |                                |                                |                                |                                | 
| SingleICMPFlow                              | ICMP     | IdleTimeout         | A -> B           | N/A              | N/A                                                                       | 1                              | 0           | Flow 1 Values                  | Flow 1 Values                  | N/A                            | N/A                            | 
| TwoICMPFlowsSameSourceInTimeout             | ICMP     | IdleTimeout         | A -> B           | A -> B           | True                                                                      | 2                              | 0           | Flow 1 Values                  | Zeroes                         | Flow 2 Values                  | Zeroes                         | 
| TwoICMPFlowsSameSourceOutOfTimeout          | ICMP     | IdleTimeout         | A -> B           | A -> B           | False                                                                     | 2                              | 0           | Flow 1 Values                  | Zeroes                         | Flow 2 Values                  | Zeroes                         | 
| TwoICMPFlowsFlippedSourceInTimeout          | ICMP     | IdleTimeout         | A -> B           | B -> A           | True                                                                      | 2                              | 0           | Flow 1 Values                  | Zeroes                         | Zeroes                         | Flow 2 Values                  | 
| TwoICMPFlowsFlippedSourceOutOfTimeout       | ICMP     | IdleTimeout         | A -> B           | B -> A           | False                                                                     | 2                              | 0           | Flow 1 Values                  | Zeroes                         | Zeroes                         | Flow 2 Values                  | 
| SingleUDPFlow                               | UDP      | IdleTimeout         | A -> B           | N/A              | N/A                                                                       | 1                              | 0           | Flow 1 Values                  | Zeroes                         | N/A                            | N/A                            | 
| TwoUDPFlowsSameSourceInTimeout              | UDP      | IdleTimeout         | A -> B           | A -> B           | True                                                                      | 1                              | 0           | Flow 1 Values + Flow 2 Values  | Zeroes                         | N/A                            | N/A                            | 
| TwoUDPFlowsSameSourceOutOfTimeout           | UDP      | IdleTimeout         | A -> B           | A -> B           | False                                                                     | 2                              | 1           | Flow 1 Values                  | Zeroes                         | Flow 2 Values                  | Zeroes                         | 
| TwoUDPFlowsFlippedSourceInTimeout           | UDP      | IdleTimeout         | A -> B           | B -> A           | True                                                                      | 1                              | 0           | Flow 1 Values                  | Flow 2 Values                  | N/A                            | N/A                            | 
| TwoUDPFlowsFlippedSourceOutOfTimeout        | UDP      | IdleTimeout         | A -> B           | B -> A           | False                                                                     | 2                              | 1           | Flow 1 Values                  | Zeroes                         | Zeroes                         | Flow 2 Values                  | 
| SingleTCPIdleOutFlow                        | TCP      | IdleTimeout         | A -> B           | N/A              | N/A                                                                       | 1                              | 0           | Flow 1 Values                  | Zeores                         | N/A                            | N/A                            | 
| TwoTCPIdleOutFlowsSameSourceInTimeout       | TCP      | IdleTimeout         | A -> B           | A -> B           | True                                                                      | 1                              | 0           | Flow 1 Values + Flow 2 Values  | Zeroes                         | N/A                            | N/A                            | 
| TwoTCPIdleOutFlowsSameSourceOutOfTimeout    | TCP      | IdleTimeout         | A -> B           | A -> B           | False                                                                     | 2                              | 1           | Flow 1 Values                  | Zeroes                         | Flow 2 Values                  | Zeroes                         | 
| TwoTCPIdleOutFlowsFlippedSourceInTimeout    | TCP      | IdleTimeout         | A -> B           | B -> A           | True                                                                      | 1                              | 0           | Flow 1 Values                  | Flow 2 Values                  | N/A                            | N/A                            | 
| TwoTCPIdleOutFlowsFlippedSourceOutOfTimeout | TCP      | IdleTimeout         | A -> B           | B -> A           | False                                                                     | 2                              | 1           | Flow 1 Values                  | Zeroes                         | Zeroes                         | Flow 2 Values                  | 
| SingleTCPEOFFlow                            | TCP      | EndOfFlow           | A -> B           | N/A              | N/A                                                                       | 1                              | 0           | Flow 1 Values                  | Zeroes                         | N/A                            | N/A                            | 
| TwoTCPEOFFlowsSameSourceInTimeout           | TCP      | EndOfFlow           | A -> B           | A -> B           | True                                                                      | 2                              | 1           | Flow 1 Values                  | Zeroes                         | Zeroes                         | Flow 2 Values                  | 
| TwoTCPEOFFlowsSameSourceOutOfTimeout        | TCP      | EndOfFlow           | A -> B           | A -> B           | False                                                                     | 2                              | 1           | Flow 1 Values                  | Zeroes                         | Zeroes                         | Flow 2 Values                  | 
| TwoTCPEOFFlowsFlippedSourceInTimeout        | TCP      | EndOfFlow           | A -> B           | B -> A           | True                                                                      | 1                              | 0           | Flow 1 Values                  | Flow 2 Values                  | N/A                            | N/A                            | 
| TwoTCPEOFFlowsFlippedSourceOutOfTimeout     | TCP      | EndOfFlow           | A -> B           | B -> A           | False                                                                     | 2                              | 1           | Flow 1 Values                  | Zeroes                         | Zeroes                         | Flow 2 Values                  |

**Comment by Zalgo2462** [comment]
Currently, `TestTwoIcmpFlowSameSource` should be renamed `TestTwoIcmpFlowSameSourceInTimeout` and `TestTwoIcmpFlowFlippedSource` should be renamed `TestTwoIcmpFlowFlippedSourceInTimeout`

**Comment by Zalgo2462** [comment]
I've added two more helper functions. In total, there is now
- `requireFlowStitchedWithZeroes(t testing.T, flow ipfix.Flow, sess *session.Aggregate)`
- `requireFlowsStitchedSameSide(t testing.T, flow1, flow2 ipfix.Flow, sessAgg *session.Aggregate)`
- `requireFlowsStitchedFlippedSides(t testing.T, flow1, flow2 ipfix.Flow, sessAgg *session.Aggregate)`

These three functions should be enough to test the outputs of the test cases above.
In the above table session aggregates are represented in pairs (AB, BA) where each component represents one side of the session. 

`requireFlowStitchedWithZeroes` is used to ensure a session aggregate matches (flow values, zeroes) or (zeroes, flow values) as shown above in the table for each aggregate (AB, BA).

`requireFlowsStitchedSameSide` is used to ensure a session aggregate matches (flow 1 values + flow 2 values, zeroes) as shown above in the table. It will also succeed if the session aggregate matches (zeroes, flow 1 values + flow 2 values).

`requireFlowsStitchedFlippedSides` is used to ensure a session aggregate matches (flow 1 values, flow 2 values) as shown above in the table. It will also succeed if the session aggregate matches (flow 2 values, flow 1 values). 

I've renamed the current tests to match the table and added two UDP tests `TestTwoUDPFlowsSameSourceInTimeout` and `TestTwoUDPFlowsFlippedSourceInTimeout` to demonstrate the use of the new functions.
